Towny/src/main/java/com/palmergames/bukkit/towny/command/TownCommand.java

```diff
@@ -4459,4 +4495,88 @@ private static int homeBlockDistance(Town town1, Town town2) {
 
 		return (int) MathUtil.distance(town1.getHomeBlockOrNull().getCoord(), town2.getHomeBlockOrNull().getCoord());
 	}
+
+	private void parseTownForSaleCommand(Player player, String[] args) throws TownyException {
+		if (args.length == 0)
+			throw new TownyException(Translatable.of("msg_error_must_be_num"));
+
+		double forSalePrice = MathUtil.getDoubleOrThrow(args[0]);
+		Town town = getTownFromPlayerOrThrow(player);
+
+		Confirmation
+			.runOnAccept(() -> {
+				setTownForSale(town, forSalePrice, false);
+				TownyMessaging.sendPrefixedTownMessage(town, Translatable.of("msg_town_forsale", town.getName(), prettyMoney(forSalePrice)));
+			})
+			.setTitle(Translatable.of("msg_town_sell_confirmation", prettyMoney(forSalePrice)))
+			.sendTo(player);
+	}
+
+	private void parseTownNotForSaleCommand(Player player) throws TownyException {
+		Town town = getTownFromPlayerOrThrow(player);
+
+		if (!town.isForSale())
+			throw new TownyException(Translatable.of("msg_town_buytown_not_forsale"));
+		
+		setTownNotForSale(town, false);
+		TownyMessaging.sendPrefixedTownMessage(town, Translatable.of("msg_town_notforsale", town.getName()));
+	}
+
+	public static void setTownForSale(Town town, double price, boolean admin) {
+		if (town != null) {
+			town.setForSale(true);
+			town.setForSalePrice(price);
+			town.save();
+		}
+	}
+
+	public static void setTownNotForSale(Town town, boolean admin) {
+		if (town != null) {
+			town.setForSale(false);
+			town.save();
+		}
+	}
+
+	private void parseTownBuyTownCommand(Player player, String[] args) throws TownyException {
+		catchConsole(player);
+
+		if (args.length == 0) {
+			throw new TownyException(Translatable.of("msg_specify_name"));
+		}
+
+		Town town = getTownOrThrow(args[0]);
+
+		if (!town.isForSale()) {
+			throw new TownyException(Translatable.of("msg_town_buytown_not_forsale"));
+		}
+		
+		Resident resident = getResidentOrThrow(player);
+		if (resident.isMayor()) {
+			throw new TownyException(Translatable.of("msg_town_buytown_already_mayor", resident.getTownOrNull().getName()));
+		}
+
+		Confirmation
+			.runOnAccept(() -> {
+				Resident currentMayor = town.getMayor();
+				
+				try {
+					if (resident.hasTown())
+						resident.removeTown();
+					townAddResident(town, resident);
+					town.setMayor(resident);
+				} catch (AlreadyRegisteredException e) {
+					town.setMayor(resident);
+				}
+				
+				TownyMessaging.sendPrefixedTownMessage(town, Translatable.of("msg_new_mayor", resident.getName()));
+				
+				currentMayor.getAccount().deposit(town.getForSalePrice(), "Payment for town sale");

```

Comment:
There is always a possibility that a town's mayor is null, you should only pay if currentMayor != null